crux/src/Crux.hs

```diff
@@ -117,53 +130,97 @@ showVersion l = outputLn ("crux version: " ++ Crux.version ++ ", " ++
     ) =>
-    OnlineBackend scope solver fs -> IO a) -> IO a
-withBackend cruxOpts nonceGen f =
-  case floatMode cruxOpts of
-    "real" -> withSolver FloatRealRepr (solver cruxOpts)
-    "ieee" -> withSolver FloatIEEERepr (solver cruxOpts)
-    "uninterpreted" -> withSolver FloatUninterpretedRepr (solver cruxOpts)
+    FloatModeRepr fm -> OnlineBackend scope solver (Flags fm) -> IO a) -> IO a
+withSelectedOnlineBackend cruxOpts nonceGen selectedSolver maybeExplicitFloatMode k =
+  case fromMaybe (floatMode cruxOpts) maybeExplicitFloatMode of
+    "real" -> withOnlineBackendFM FloatRealRepr
+    "ieee" -> withOnlineBackendFM FloatIEEERepr
+    "uninterpreted" -> withOnlineBackendFM FloatUninterpretedRepr
     "default" ->
-      case (solver cruxOpts) of
-        s@"cvc4" -> withSolver FloatRealRepr s
-        s@"yices" -> withSolver FloatRealRepr s
-        s@"z3" -> withSolver FloatIEEERepr s
-        s -> fail $ "unknown solver: " ++ s ++ "; expepected of one [cvc4|yices|z3]."
-    fm -> fail $ "unknown floating-point mode: " ++ fm ++ "; expepected of one [real|ieee|uninterpreted|default]."
+      case selectedSolver of
+        CCS.Yices -> withOnlineBackendFM FloatRealRepr
+        CCS.CVC4 -> withOnlineBackendFM FloatRealRepr
+        CCS.STP -> withOnlineBackendFM FloatRealRepr
+        CCS.Z3 -> withOnlineBackendFM FloatIEEERepr
+    fm -> fail ("Unknown floating point mode: " ++ fm ++ "; expected one of [real|ieee|uninterpreted|default]")
   where
-  unsatCores | yicesMCSat cruxOpts = NoUnsatFeatures
-             | otherwise           = ProduceUnsatCores
-  withSolver fpMode "cvc4" =
-    withCVC4OnlineBackend fpMode nonceGen ProduceUnsatCores $ \sym->
-      do case goalTimeout cruxOpts of
-           Just s -> symCfg sym cvc4Timeout (floor (s * 1000))
-           Nothing -> return ()
-         f sym
-  withSolver fpMode "yices" =
-    withYicesOnlineBackend fpMode nonceGen unsatCores $ \sym ->
-      do symCfg sym yicesEnableMCSat (yicesMCSat cruxOpts)
-         case goalTimeout cruxOpts of
-           Just s -> symCfg sym yicesGoalTimeout (floor s)
-           Nothing -> return ()
-         f sym
-  withSolver fpMode "z3" =
-    withZ3OnlineBackend fpMode nonceGen ProduceUnsatCores $ \sym->
-      do case goalTimeout cruxOpts of
-           Just s -> symCfg sym z3Timeout (floor (s * 1000))
-           Nothing -> return ()
-         f sym
-  withSolver _fpMode other =
-    fail $ "unknown solver: " ++ other ++ "; expepected of one [cvc4|yices|z3]."
+    unsatCores | yicesMCSat cruxOpts = NoUnsatFeatures
+               | otherwise           = ProduceUnsatCores
+
+    withOnlineBackendFM floatRepr =
+      case selectedSolver of
+        CCS.Yices -> withYicesOnlineBackend floatRepr nonceGen unsatCores $ \sym -> do
+          symCfg sym yicesEnableMCSat (yicesMCSat cruxOpts)
+          case goalTimeout cruxOpts of
+            Just s -> symCfg sym yicesGoalTimeout (floor s)
+            Nothing -> return ()
```

Comment:
I wish we had a way to reduce some of the duplication related to setting timeouts here. It makes me wonder whether a more global setting, like `solverInteractionFile`, would be appropriate. It could have a fixed set of units (seconds?) and solvers would internally have to calculate their own timeouts based on that.

Reply:
Oh my yes, that would be nice.